File: Java/src/NapakalakiGame/SpecificBadConsequence.java
/*
 * Autores:
 * Guillermo Montes Martos
 * Manuel Ruiz López
 *
 */
package NapakalakiGame;

import java.util.ArrayList;

public class SpecificBadConsequence extends BadConsequence{
    private ArrayList<TreasureKind> specificVisibleTreasures = new ArrayList();
    private ArrayList<TreasureKind> specificHiddenTreasures = new ArrayList();
    
    public SpecificBadConsequence (String t, int l, ArrayList<TreasureKind> v, ArrayList<TreasureKind> h){
        super(t, l);
        specificVisibleTreasures = v;
        specificHiddenTreasures = h;
    }
     
    public ArrayList<TreasureKind> getSpecificVisibleTreasures() {
        return specificVisibleTreasures;
    }
    
    public ArrayList<TreasureKind> getSpecificHiddenTreasures() {
        return specificHiddenTreasures;
    }

    @Override
    public boolean isEmpty() {
    return specificVisibleTreasures.isEmpty() && specificHiddenTreasures.isEmpty();
    }
    
    @Override
    public void substractVisibleTreasure(Treasure t) {
        specificVisibleTreasures.remove(t.getType());
    }
    
    @Override
    public void substractHiddenTreasure(Treasure t) {
        specificHiddenTreasures.remove(t.getType());
    }
    
    @Override
    public BadConsequence adjustToFitTreasureLists(ArrayList<Treasure> v, ArrayList<Treasure> h) {
        BadConsequence nuevo;

        ArrayList<TreasureKind> sv, sh;
        sv = intersection(specificVisibleTreasures, v);
        sh = intersection(specificHiddenTreasures, h);
        nuevo = new SpecificBadConsequence(this.getText(), this.getLevels(), sv, sh);
        
        return nuevo;
    }
    
    private ArrayList<TreasureKind> intersection(ArrayList<TreasureKind> tk, ArrayList<Treasure> t) {
    ArrayList<TreasureKind> salida = new ArrayList();

    if(!tk.isEmpty() && !t.isEmpty())
    {
        ArrayList<TreasureKind> aux = new ArrayList(tk);
        
        for(Treasure treasure:t) {
            int i = aux.indexOf(treasure.getType());
            if(i != -1)
                salida.add(aux.remove(i));
        }
    }

    return salida;
    }
}

File: Ruby/lib/SpecificBadConsequence.rb
#encoding: utf-8

# Autores:
# 
#     Guillermo Montes Martos
#     Manuel Ruiz López
#

require_relative 'BadConsequence.rb'
require_relative 'TreasureKind.rb'


module NapakalakiGame
  class SpecificBadConsequence < BadConsequence
    
    attr_reader :specificVisibleTreasures
    attr_reader :specificHiddenTreasures
    
    public_class_method:new
    
    def initialize(t, l, sVisible, sHidden)
      super(t, l)
      @specificVisibleTreasures = sVisible
      @specificHiddenTreasures = sHidden
    end
  
    def isEmpty
      @specificVisibleTreasures.empty? && @specificHiddenTreasures.empty?
    end
    
    def substractVisibleTreasure(t)
      if @specificVisibleTreasures.size > 0
        @specificVisibleTreasures.delete_at(@specificVisibleTreasures.index(t.type))
      end
    end
    
    def substractHiddenTreasure(t)
      if @specificHiddenTreasures.size > 0
        @specificHiddenTreasures.delete_at(@specificHiddenTreasures.index(t.type))
      end
    end
    
    def adjustToFitTreasureLists(v, h)
      bd = nil
      interV = intersection(@specificVisibleTreasures, v) 
      interH = intersection(@specificHiddenTreasures, h)
      bd = SpecificBadConsequence.new(@text,@levels,interV,interH)
    end
    
    private
    def intersection(arrayTK, arrayTreasure)  # necesita ambos vectores para crear un método general
      salida = Array.new
      
      if !arrayTK.empty? && !arrayTreasure.empty?
        aux = arrayTK.clone   # Copia vector TK de BadConsequence
        for elemento in arrayTreasure
          unless (i = aux.index(elemento.type)) == nil
            salida << aux.delete_at(i)
          end 
        end
      end
      
      salida
    end
    
  end
end